README.md
**OpenESS 0.86.0**   is a network-capable sound server libary mainly for embedded systems and other operatins systems.
OpenESS is free and open-source software, and is licensed under the terms of the GNU Lesser General Public License.

<a href="https://www.codacy.com/app/RoseLeBlood/openess?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=RoseLeBlood/openess&amp;utm_campaign=Badge_Grade"><img src="https://api.codacy.com/project/badge/Grade/4f0ba2c68a904b8da2f1d45d5f3596d4"/></a>
<img src="https://img.shields.io/github/license/RoseLeBlood/Openess.svg"/>
<img src="https://img.shields.io/github/last-commit/roseleblood/openess.svg"/>
<a href="https://twitter.com/padersophia">
<img src="https://img.shields.io/twitter/follow/padersophia.svg"></a>


_For informations, examples and usage, please refer to the [Wiki][wiki]_

## Current Features

* Powerful core that allows for easy to port
* Powerful socket abscrations layer (SAL) only dram and dramlite at the moment
* Easy to use platform configuration
* mutex, spinlock, task and ringbuffer handling on various platform (current only esp32)
* generic backends:  i2s, more work in progress
* task are multitaskin safe
* code style min B
* C++ System

## Progressed features (when ready than leave beta phase)

* running example server on esp32 and linux
* MQTT status upport and logging (abscrations layer)
* Mini Webserver  - TCP for status (Port 8080)

## Planned features

* Improve documentation

* **Suggestions?** Feel free to create an issue and tag it with feature request.


## Release History
See [CHANGELOG.md][changelog]



[wiki]: https://github.com/RoseLeBlood/openess/wiki
[changelog]:CHANGELOG.md


CHANGELOG.md

## Release History
* 0.86.0:
  - add: singleton object class,  lockable_object class, task_group class, ess_wifi_net_device interface
  - add esp32_ap_wifi class
  - ess_audioblock_t allocater remove
  - module and channel update read function: from  ess_audioblock_t* to ess_audioblock_t&
* 0.85.1:
  - add lib/tiny-http libary
  - add is_equels to ip4 and ip6 address class
  - add ess_stream and ess_stream_writer class
  - rename ess_audio_buffer to ess_audio_buffer_stream
  - add ESS_CONFIG_STREAM_WRITER_GENERIC config to all platform defines
* 0.85.0:
  - remove old socket system
  - add ess_ipaddress , ess_ip4address, ess_ip6address class
  - add ess_endpoint, ess_ip_endpoint4_t, ess_ip_endpoint6_t
  - add ess_sockets#: inet, inet_stream - ip4/ip6
* 0.8.21:
  - reorder header
  - rename ess_inout_module to ess_effect
  - new base from ess_effect is ess_input_module
  - add ess_dsp_multiply - simple DSP with 2 Inputs and 1 Output (InA * InB = Out)
  - add ess_dsp_multiplexer - simple DSP with 1 Inputs and 2 Output (OutB = In; OutA = In)
  - add ess_channel: virtual bool is_connected
* 0.8.11:
  - call ess_audio_memory_map in ess_platform creating - bug fix
  - update ref implatierung esp32 debug output
  - add ess_debug.h helper functions and example for  ess_output_analyzed_module and the configs ESS_PLATFORM_MONTORING, ESS_OUTPUT_TIME_ANALYZED
* 0.8.10:
  - add ess_error_t ess_task::start(uint32_t core)
  - add ess_audio_block_t
  - update read functions in channel and modules virtual unsigned int  read(ess_audioblock_t *block, unsigned int offset)
  - add memory map for ess_audio_block_t
  - add ess_output_analyzed_module - output module with time analyzed
  - add new configs and reoder:
    - ESS_MAX_AUDIO_MEMORY 163840,
    - ESS_MEMORY_MAP_DEBUG
    - ESS_PLATFORM_MONTORING 1 /**< when set then useble get_cpu_max and get_cpu_load*/
    - ESS_OUTPUT_TIME_ANALYZED
* 0.8.01:
  - add  class ess_stereo_simple_buffer_output_module
    mix floated audio data to a simple integer buffer stereo
  - ess_esp32i2s_output_module changes:
    basic class ess_stereo_simple_buffer_output_module
    ess_esp32uart_output_module changes:
      basic class ess_stereo_simple_buffer_output_module
  - add class ess_system_format:
    static class holder for formart handling
  - ess_platform_esp32 remove:
    get_bits, get_samplerate, get_channels functions and move it to ess_system_format
  - ess_format add:
      ess_format_from_string(const char*)
      remove issus
  - ess_i2s_controller esp32:
      remove issus
      get_format
  - ess_format_t add:
      2.1, Quad , 5.1, 7.1 Formats
* 0.8.00:
  - add udplite output module
  - start esp32 refecrence implantierung (private git - public in the future)
  - version update to 0.8
* 0.6.04:
  - remove TeensyAudioLibrary
  - add very flexible audio system
  - rename class see examples
  - add ess_channel
  - add connecting system
  - fix issus
* 0.5.0:
  - update to 0.5
  - start impl new frontend system. based on the TeensyAudioLibrary with modifications
  - rename backends:
      - i2s_generic_backend -> i2s_generic_output_backend
      - generic_openal_backend -> generic_openal_output_backend
      - generic_udp_backend -> generic_udp_output_backend
* 0.4.5:
  - add IP4/IP6 multicast
  - add new system
  - remove ess_context
* 0.4.2:
  - remove issus
* 0.4.03
  - add ess_backend::get_blksize
  - add config ESS_BUF_COUNT 4
  - add config ESS_BUF_SIZE      1024
* 0.4.02
  - add ess_network.h for socket multiplatform layer
  - update example
  - ess_dram_server using config from config.h
  - add generic_udp_backend
* 0.4.0
  - add udp_inet server / client backend
  - add ess_server and ess_dram_server (basic)
  - add esp32 dram server example
* 0.3.4
  - add backend factory
  - add context creating functions
  - add probe function to ess_backend as virtual
  - add esp32 i2s_generic_backend
* 0.3
  - update platform abstraction layer
  - add mutex, task and spinlock functions - task using mutex on esp32 platforms
  - add examples
* 0.2
  - rename headers
  - add platform abstraction layer
* 0.0.1
  - Work in progress


include/config.h
/****************************************************************************
 * Copyright (C) 2019 by Anna Sopdia Schröck                                *
 *                                                                          *
 * This file is part of ess.                                                *
 *                                                                          *
 *   ess is free software: you can redistribute it and/or modify it         *
 *   under the terms of the GNU Lesser General Public License as published  *
 *   by the Free Software Foundation, either version 3 of the License, or   *
 *   (at your option) any later version.                                    *
 *                                                                          *
 *   ess is distributed in the hope that it will be useful,                 *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of         *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the          *
 *   GNU Lesser General Public License for more details.                    *
 *                                                                          *
 *   You should have received a copy of the GNU Lesser General Public       *
 *   License along with Box.  If not, see <http://www.gnu.org/licenses/>.   *
 ****************************************************************************/

/**
 * @file config.h
 * @author Anna Sopdia Schröck
 * @date 30 Januar 2019
 * @brief File containing configurations
 *
 */

#ifndef __ESS_CONFIG_H__
#define __ESS_CONFIG_H__

#define ESS_ON 1
#define ESS_OFF 0
#define ESS_PROTOCOL_UDP							1
#define ESS_PROTOCOL_TCP							2
#define ESS_PROTOCOL_UDP_LITE		3
#define ESS_FAMILY_IP4											4
#define ESS_FAMILY_IP6											6
#define ESS_FAMILY_BOTH 							46
// ----------------------------- main config ----------------
#define ESS_DEFAULT_AUDIO_PACKET_SIZE 1024
#define ESS_DEFAULT_SERVER_FORMAT ESS_FORMAT_STEREO_44100_16
#define ESS_DEFAULT_SERVER_PACKET_SIZE ESS_DEFAULT_AUDIO_PACKET_SIZE
// ----------------------------- ess_audio_memory_map config ----------------
#define ESS_MEMORY_MAP_DEBUG ESS_ON
#define ESS_MEMORY_MAP_EXTERN ESS_OFF
#define ESS_MAX_AUDIO_MEMORY 163840
#define ESS_MAX_AUDIO_BLOCKS   64
#define ESS_MOD_AUDIO_BLOCKS   32
#define ESS_MIN_AUDIO_BLOCKS 		16
#define ESS_USED_AUDIO_BLOCKS ESS_MOD_AUDIO_BLOCKS
// ----------------------------- ess_http_request server config ----------------
#define ESS_CONFIG_HTTP_REQUEST_SERVER ESS_ON
#define ESS_CONFIG_HTTP_REQUEST_MAX_HEADERS 32
#define ESS_CONFIG_HTTP_REQUEST_MAX_ARGS 32

#define ESS_CONFIG_HTTP_REQUEST_PORT 8686
#define ESS_CONFIG_HTTP_REQUEST_FAMILY ESS_FAMILY_IP4
// ----------------------------- ess_debug config ----------------
#define ESS_PLATFORM_MONTORING ESS_OFF /**< when set then useble get_cpu_max and get_cpu_load (experimental)*/
#define ESS_TIME_ANALYZED_MODULE ESS_ON
#define ESS_CHANNEL_DEBUG ESS_ON
#define ESS_MODULE_DEBUG ESS_ON

#if ESS_CHANNEL_DEBUG == ESS_ON
#define ESS_CHANNEL_OUTPUT_DEBUG ESS_ON
#define ESS_CHANNEL_INPUT_DEBUG ESS_ON
#else
#define ESS_CHANNEL_OUTPUT_DEBUG ESS_OFF
#define ESS_CHANNEL_INPUT_DEBUG ESS_OFF
#endif
// ---------------------------- ess_platform configs ------------
#define ESS_PLATFORM_ESP32					ESS_ON/** @brief If defined compiled backend for esp32 */
#define ESS_PLATFORM_RPI 					 		 ESS_OFF/** @brief If defined compiled backend for Raspberry PI (linux) */
#define ESS_PLATFORM_LINUX   			ESS_OFF/** @brief If defined compiled backend for generic linux (openal) */
#define ESS_PLATFORM_WINDOWS  ESS_OFF/** @brief If defined compiled backend for  generic windows (openal, wasapi) */

#if  ESS_PLATFORM_ESP32 == ESS_ON
	#include "freertos/FreeRTOS.h"
	#define ESS_CONFIG_NETWORK_ESP32 /**< esp32 using own network functions */
	#define ESS_CONFIC_TASK_ESP32	/**< esp32 using own task functions */
	#define ESS_CONFIG_RINGBUFFER_ESP32 /**< esp32 using own ringbuffer functions */
	#define ESS_CONFIG_MUTEX_ESP32 /**< esp32 using own mutex functions */
	#define ESS_CONFIG_SPINLOCK_ESP32 /**< esp32 using own spinlock functions */
	#define ESS_CONFIG_MEMORY_ESP32 /**< esp32 using own memory functions */
	#define ESS_CONFIG_STREAM_WRITER_GENERIC

	#define ESS_ENABLE_BACKEND_OUT_I2S /**< esp32 platform I2S backend available */
	//#define ESS_ENABLE_OUTMODULE_UDPLITE/**< generic platform UDP backend available */
	#define ESS_ENABLE_OUTMODULE_UART
	#define ESS_DEFAULT_SERVER_NAME "OpenESS-esp32" /**< basic server name*/
	#define ESS_IRAM_ATTR IRAM_ATTR
	#define ESS_CONFIC_MAX_CORES 2
#endif // ESS_PLATFORM_ESP32

#if ESS_PLATFORM_RPI == ESS_ON
	#define ESS_CONFIG_MUTEX_GENERIC
	#define ESS_CONFIC_TASK_GENERIC
	#define ESS_CONFIG_RINGBUFFER_GENERIC
	#define ESS_CONFIG_MEMORY_GENERIC
	#define ESS_CONFIG_STREAM_WRITER_GENERIC
	/** @brief If defined then OpenAL backend available */
	#define ESS_ENABLE_BACKEND_OUT_OPENAL
	/** @brief If defined then UDP backend available */
	#define ESS_ENABLE_BACKEND_OUT_UDP
	#define ESS_DEFAULT_SERVER_NAME "OpenESS-rpi"
	#define ESS_IRAM_ATTR IRAM_ATTR
	#define ESS_CONFIC_MAX_CORES 1
#endif //ESS_PLATFORM_RPI

#if ESS_PLATFORM_LINUX == ESS_ON
	#define ESS_CONFIC_TASK_GENERIC
	#define ESS_CONFIG_RINGBUFFER_GENERIC
	#define ESS_CONFIG_MUTEX_GENERIC
	#define ESS_CONFIG_MEMORY_GENERIC
	#define ESS_CONFIG_STREAM_WRITER_GENERIC
	/** @brief If defined then OpenAL backend available */
	#define ESS_ENABLE_BACKEND_OUT_OPENAL
	/** @brief If defined then UDP backend available */
	#define ESS_ENABLE_BACKEND_OUT_UDP
	#define ESS_DEFAULT_SERVER_NAME "OpenESS-linux"
	#define ESS_CONFIC_MAX_CORES 8
	#define ESS_IRAM_ATTR /* */
#endif //ESS_PLATFORM_LINUX

#if ESS_PLATFORM_WINDOWS == ESS_ON
	#define ESS_CONFIG_SEMAPHORE_WINDOWS
	#define ESS_CONFIC_TASK_WINDOWS
	#define ESS_CONFIG_RINGBUFFER_WINDOWS
	#define ESS_CONFIG_MUTEX_WINDOWS
	#define ESS_CONFIG_MEMORY_GENERIC
	#define ESS_CONFIG_STREAM_WRITER_GENERIC
	/** @brief If defined then OpenAL backend available */
	#define ESS_ENABLE_BACKEND_OUT_OPENAL
	/** @brief If defined then UDP backend available */
	#define ESS_ENABLE_BACKEND_OUT_UDP
	#define ESS_DEFAULT_SERVER_NAME "OpenESS-windows"
	#define ESS_CONFIC_MAX_CORES 8
	#define ESS_IRAM_ATTR /* */
#endif //ESS_PLATFORM_WINDOWS
//-------------------------------------------------------------------------------------------------------
#ifdef ESS_ENABLE_OUTMODULE_UDPLITE
	#define ESS_OUT_UDPLITE_SENDTO_PORT 8086
	#define ESS_OUT_UDPLITE_OUTPUT_CHANNELS ESS_CHANNEL_FORMAT_STEREO // 2

	#if ESS_DEFAULT_SERVER_FAMILY == ESS_FAMILY_IP4
	#define ESS_OUT_UDP_SENDTO_HOST "192.168.0.235"
	#else
	#define ESS_OUT_UDP_SENDTO_HOST "::2"
	#endif
#endif

#ifdef ESS_ENABLE_BACKEND_OUT_I2S
	#define I2S_EXTERNAL_DAC_BCK 26
	#define I2S_EXTERNAL_DAC_LRCLK 25
	#define I2S_EXTERNAL_DAC_DOUT 22
	#define  I2S_EXTERNAL_DAC_DIN   23

	#define ESS_BACKEND_I2S_DMA_BUF_SIZE	ESS_DEFAULT_AUDIO_PACKET_SIZE
	#define ESS_BACKEND_I2S_DMA_BUF_COUNT 3
#endif



#endif
